### ISSUE TITLE
What triggers a replace?  [PFC+keys dupe]

### ISSUE THREAD

**Issue by electerious** [developit/preact#517]
I replaced React with Preact for testing purposes. My site uses SSR and executes the same JS on the client. React shows a handy warning when the HTML from the server is different to the initial rendering on the client. The warning includes the difference and makes it easy to find the reason for it. Preact however replaces the whole HTML and I'm not sure why. I'm now looking for a way to find the reason.

Server:

```js
`<div id="preact">${ render(<Content { ...props } /> }</div>`
```

Client:

```js
const root = document.querySelector('#preact')

render(<Content { ...props } />, root, root.lastChild)
```

What I know:

- Server and client have the same props.
- React was able to reuse HTML rendered on the server
- preact v7.1.0
- preact-render-to-string v3.3.0

Video:

- https://cl.ly/iod7
- I delayed the client rendering for 3 seconds
- First green flash => Initial paint
- Second green flash => Repaint after client rendering triggers

Here's the [React version of the site](https://laudableapps.com).

Let me know if more details are necessary.

**Comment by developit** [comment]
This looks like the issue where keys and functional components cause diff reflow/mismatch.  Are you by any chance using keys + functional components?  If so, that is being fixed thanks to a rewrite of functional component support in `7.2.0` - you can see the change [here](https://github.com/developit/preact/commit/ca478e325c39834a6201a7f0adb588a53785a5d8).

One other thing to try: instead of `root.lastChild`, try `root.lastElementChild`.

**Comment by electerious** [comment]
Yes, I'm using keys and functional components. I've downloaded the current master, built it and placed it into `node_modules/preact`. The problem still occurs. Even when using `root.lastElementChild`.

**Comment by developit** [comment]
Update: this is caused by keys on Pure Functional Components.  It is fixed by [the rewrite of functional components](https://github.com/developit/preact/commit/ca478e325c39834a6201a7f0adb588a53785a5d8), which will be release in `8.0` (hopefully quite soon!).

**Comment by developit** [comment]
This is finally fixed in 8.0 💃
